subsystems/mvdm/ntvdm/bios/vidbios.c

```diff
@@ -2942,6 +2942,303 @@ static VOID VidBiosPrintCharacter(CHAR Character, BYTE Attribute, BOOLEAN UseAtt
+{
+    if (VideoMode > BIOS_MAX_VIDEO_MODE)
+        return 0;
+    return 0x10000 / VideoModes[VideoMode].PageSize;
+}
+
+static VOID VidBiosSetPixel(WORD X, WORD Y, BYTE Page, BYTE Value)
+{
+    SHORT ScreenWidth = VgaGetDisplayResolution().X;
+
+    if (IS_TEXT_MODE(Bda->VideoMode))
+    {
+        DPRINT1("Note, trying to modify pixel (%d,%d) @ page %d to %d in text mode %X - returning 0.\n", X, Y, Page, Value, Bda->VideoMode);
+        return;
+    }
+
+    switch (Bda->VideoMode)
+    {
+        case 0x04: /* CGA 320x200 4c */
+        case 0x05: /* CGA 320x200 4c */
+        case 0x06: /* CGA 640x200 2c  */
+        {
+            WORD CgaSegment[] = { CGA_EVEN_VIDEO_SEG, CGA_ODD_VIDEO_SEG };
+            ULONG Address;
+            BYTE MemoryValue;
+            BOOL Xor;
+            BYTE MaxColor;
+            BYTE ValueShift;
+            BYTE PixelsPerByte;
+
+            if (Bda->VideoMode == 0x06)
+            {
+                MaxColor = 1;
+                ValueShift = ~X & 7;
+                PixelsPerByte = 8;
+            }
+            else
+            {
+                MaxColor = 3;
+                ValueShift = (~X & 3) * 2;
+                PixelsPerByte = 4;
+            }
+
+            Xor = (Value & 0x80) != 0;
+            Value &= MaxColor;
+            Address = TO_LINEAR(CgaSegment[Y & 1],
+                                Page * Bda->VideoPageSize + ((Y / 2) * ScreenWidth + X) / PixelsPerByte);
+
+            EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            if (Xor)
+            {
+                MemoryValue ^= Value << ValueShift;
+            }
+            else
+            {
+                MemoryValue &= ~(MaxColor << ValueShift);
+                MemoryValue |= Value << ValueShift;
+            }
+            EmulatorWriteMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            break;
+        }
+
+        case 0x0D: /* EGA 320x200 16c */
+        case 0x0E: /* EGA 640x200 16c */
+        case 0x10: /* EGA 640x350 16c */
+        case 0x12: /* VGA 640x480 16c */
+        {
+            ULONG Address;
+            BOOL Xor;
+            BYTE OldMode;
+            BYTE OldRotation;
+            BYTE OldRead;
+            BYTE OldMask;
+            BYTE CurrentPlaneMask;
+            BYTE CurrentPlane;
+            BYTE Shift = ~X & 7;
+
+            Xor = (Value & 0x80) != 0;
+            Value &= 15;
+            Address = TO_LINEAR(GRAPHICS_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + (Y * ScreenWidth + X) / 8);
+
+            /* Make sure we're in write mode 0 */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_MODE_REG);
+            OldMode = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, 0x00);
+
+            /* Setup the rotation */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_ROTATE_REG);
+            OldRotation = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, Xor ? 0x18 : 0x00);
```

Comment:
Since here we're setting the Logical Operation field (as we should!), the whole code should use that instead of trying to emulate it.